### ISSUE TITLE
GraphQL Pro's validation does recognize the schema

### ISSUE THREAD

**Issue by waissbluth** [rmosolgo/graphql-ruby#4697]
**Describe the bug**

`GraphQL::Pro::OperationStore::Validate` if failing for valid queries (actually, failing for all our queries) with e.g, `Field 'emailAddressDeletion' doesn't exist on type 'Mutation'`.

**Versions**

`graphql` version: 1.12.19
`graphql-pro` 1.24.10
`rails` : 6.1.7.6

**Steps to reproduce**

Attempt to import valid queries via `/graphql/sync` or  `GraphQL::Pro::OperationStore::AddOperationBatch` directly

**Expected behavior**

For all queries to be added

**Actual behavior**

All queries are marked as invalid, and the validator thinks the schema is essentially empty. However, the schema is able to execute all queries just fine.

</details>

**Additional context**

Skipping validation inserts the queries to the registry properly, and they are executing with no problem. I.e.,

```

module GraphQL
  module Pro
    class OperationStore
      module Validate
        (class << self; self; end).instance_eval do
          # Skipping validation since it is failing to find all fields.
          old_validate = instance_method(:validate)
          define_method(:validate) do |*args, **kw|
            _errors, query = old_validate.bind_call(self, *args, **kw)
            [[], query]
          end
        end
      end
    end
  end
end
```

**Comment by waissbluth** [comment]
If there is no fix for this, we could also perhaps? make this work with a new argument for `AddOperationBatch` to `disable_validation`

**Comment by rmosolgo** [comment]
Hey! Sorry for the trouble. This is definitely the first time I've heard of this problem... When you describe "mutations registered dynamically," are you describing how your schema is defined in Ruby? If so, could you share an example of how it's defined?

**Comment by waissbluth** [comment]
> Hey! Sorry for the trouble. This is definitely the first time I've heard of this problem... When you describe "mutations registered dynamically," are you describing how your schema is defined in Ruby? If so, could you share an example of how it's defined?

Hi @rmosolgo, thanks for the quick reply. I just renamed the issue since I now think that the schema is not being recognized at all by the Validator (not just the metaprogramming bits).

In any case, we are doing something like this. It all works perfectly with other parts of the app (exporting the schema, executing queries, etc)

```
    mutation_class =
      Class.new(Types::BaseMutation) do
        mutation[:arguments].each do |args|
          name, type, kwargs = args
          argument(name, type, **kwargs)
        end
        type(response_class)

        define_method(:resolve) do |**kwargs|
          # body of the resolver
        end
        # ...
      end
    Object.const_set(mutation_name, mutation_class)
    Types::MutationType.field(graphql_name.underscore, mutation: mutation_class)
```

**Comment by rmosolgo** [comment]
Thanks for sharing that detail. I tried to write a replication case with that version of GraphQL-Ruby, but it gave me the same validation error, just trying to execute a query: 

```ruby
require "bundler/inline"

gemfile do
  gem "graphql", "1.12.19"
end

mutation_class = Class.new(GraphQL::Schema::RelayClassicMutation) do
  graphql_name "DoSomething"

  field :value, Integer, null:  false

  def resolve
    { value: 1 }
  end
end

class MySchema < GraphQL::Schema
  class Mutation < GraphQL::Schema::Object
    field :dummy, Boolean, null: false
  end
  mutation(Mutation)
end

MySchema::Mutation.field("doSomething", mutation: mutation_class)

res = MySchema.execute("mutation { doSomething(input: {}) { value } }")
pp res.to_h
# {"errors"=>
#   [{"message"=>"Field 'doSomething' doesn't exist on type 'Mutation'",
#     "locations"=>[{"line"=>1, "column"=>12}],
#     "path"=>["mutation", "doSomething"],
#     "extensions"=>{"code"=>"undefinedField", "typeName"=>"Mutation", "fieldName"=>"doSomething"}}]}
```

I think this is because, back in v1.12.19, GraphQL-Ruby made a backwards-compatible, cached version of each schema member (type, field, argument, etc) during schema boot. Those cached versions were used at runtime and didn't reflect changes made to definition _classes_ after the cached versions were created. 

To confirm this, I was able to make my script work by adding the field _before_ calling `mutation(...)`: 

```ruby
class MySchema < GraphQL::Schema
  class Mutation < GraphQL::Schema::Object
    field :dummy, Boolean, null: false
  end

  mutation_class = Class.new(GraphQL::Schema::RelayClassicMutation) do
    graphql_name "DoSomething"

    field :value, Integer, null:  false

    def resolve
      { value: 1 }
    end
  end
  MySchema::Mutation.field("doSomething", mutation: mutation_class)

  mutation(Mutation)
end

# Then I get the expected output: 
# {"data"=>{"doSomething"=>{"value"=>1}}}
```

So, my guess is that adding the OperationStore has changed how your schema boots, so that those (stale) caches are used by the OperationStore. I honestly don't know why execution still works fine -- I would have _expected_ execution to have been messed up to (as in my script above). 

I'm going to keep working with my replication script to see if I can find a setup where OperationStore causes an otherwise _working_ schema to become broken. 

In the meantime, could you share a few more details about your setup: 

- What causes those dynamic calls to `MutationType.field(...)`? Are those calls inside the schema definition itself, or in other files? Could you check to see if those calls are made _before_ or _after_ the OperationStore is attached to the schema? 
- Does `use GraphQL::Pro::OperationStore ...` come before or after `mutation(...)`? If it's currently before, you could try moving it to after.

Hopefully those will yield a few more clues.

**Comment by rmosolgo** [comment]
I was able to make a basically _working_ script, using this old GraphQL-Ruby version and the OperationStore (and your patch from #4696): 

```ruby
require "bundler/inline"

gemfile do
  gem "graphql", "1.12.19"
  gem "graphql-pro", "1.24.10"
  gem "redis"
end

require "logger"

module GraphQL
  class Schema
    class Member
      module HasArguments
        old_get_argument = instance_method(:get_argument)

        # Version 1.12.19 of graphql-ruby takes one argument
        define_method(:get_argument) do |arg_defn, _context = nil|
          old_get_argument.bind_call(self, arg_defn)
        end
      end
    end
  end
end

MutationClass = Class.new(GraphQL::Schema::RelayClassicMutation) do
  graphql_name "DoSomething"

  field :value, Integer, null:  false

  def resolve
    { value: 1 }
  end
end

class MySchema < GraphQL::Schema
  class Mutation < GraphQL::Schema::Object
    field :dummy, Boolean, null: false
  end
  use GraphQL::Pro::OperationStore, redis: Redis.new

  MySchema::Mutation.field("doSomething", mutation: MutationClass)
  mutation(Mutation)
end

query_str = "mutation DoSomething { doSomething(input: {}) { value } }"
res = MySchema.execute(query_str)
pp res.to_h
# {"data"=>{"doSomething"=>{"value"=>1}}}


MySchema.operation_store.upsert_client("client-1", "secret-token")
MySchema.operation_store.add(body: query_str, client_name: "client-1", operation_alias: "1234")
res2 = MySchema.execute(nil, context: { operation_id: "client-1/1234"})
pp res2.to_h
# {"data"=>{"doSomething"=>{"value"=>1}}}
``` 

So, I think the problem must somehow arise from how the schema is built in Ruby-land. Let me know about those questions above!

**Comment by waissbluth** [comment]
Thanks for the investigation @rmosolgo!

> To confirm this, I was able to make my script work by adding the field before calling mutation(...):
Yes, that is exactly what we do.

To answer your questions:

> What causes those dynamic calls to MutationType.field(...)? Are those calls inside the schema definition itself, or in other files? Could you check to see if those calls are made before or after the OperationStore is attached to the schema?

We essentially do a `require(filename)` of files that then run register the mutations (one file per mutation). Then we `mutation(Types::MutationType)` inside the schema.

> Does use GraphQL::Pro::OperationStore ... come before or after mutation(...)? If it's currently before, you could try moving it to after.

It is after already. We had it before and did not make a difference.

Question for you: in your example does `OperationStore::Validate.validate` succeed?

**Comment by rmosolgo** [comment]
Yes, stand-alone validation works, too. Here's what I added to my script: 

```ruby
pp GraphQL::Pro::OperationStore::Validate.validate(MySchema, GraphQL.parse(query_str), client_name: "client-1")
# [[], query ...]
#  ^^ An empty array of errors
```

Hmm, thanks for sharing that debugging info. Nothing there jumps out at me. Could you just confirm that the operation store has a reference to the right kind of schema object, by putting this in a Rails console (or whatever): 

```
pp [MySchema.operation_store.schema, MySchema.operation_store.schema.class]
pp MySchema.operation_store.schema.get_field("Mutation", "emailAddressDeletion")
```

(Use your schema instead of `MySchema`.) 

I'd expect that to print out `[MySchema, Class]` and the `emailAddressDeletion` field definition. 

One more thing: I see your example code uses `graphql_name.underscore`, but `emailAddressDeletion` is camel case. Does `email_address_deletion` work better? 

Let me know what you find!

**Comment by waissbluth** [comment]
> I'd expect that to print out [MySchema, Class] and the emailAddressDeletion field definition.

It does print out `[MySchema, Class]`, but the second one shows `nil`, which I guess is consistent with the validation error we are getting.

Can you provide a non-OperationStore version of that `get_field`, to confirm the field is there?

**Comment by rmosolgo** [comment]
Sure, a non-operation-store version would be: 

```ruby
MySchema.get_field("Mutation", "emailAddressDeletion")
```

Does that return a `GraphQL::Schema::Field` instance? Another way would be: 

```ruby 
Types::MutationType.get_field("emailAddressDeletion")
```

**Comment by waissbluth** [comment]
wow yeah all of those are nil. All mutations defined from outside the `Types::MutationType` class are not found in the schema (i.e., via `Types::MutationType.field(graphql_name.underscore, mutation: mutation_class)`). However, the app does work fine.

I did some digging and it turns out I can get the app to fail with the same error if I stop passing in `current_user` into the `context` object that we are passing in to the `execute

**Comment by rmosolgo** [comment]
Oh, interesting! Maybe the operation store was the first time that the app had tried using GraphQL _without_ a `current_user`. I don't know about `access_level` constraints, but if you could share the code that causes those GraphQL definition files to be `require`d, and the code that implements `access_level`, maybe we could find some clues.

**Comment by waissbluth** [comment]
Oh i thought access level was a graphql-ruby thing but this is getting a bit app-specific, so it turns out I think we could work around this if we were able to pass in a `context` object to `OperationStore::AddOperationBatch.call` that would need to be passed in all the way through to the call to `validate`. Does that make sense?

Thanks!

**Comment by rmosolgo** [comment]
Yes, that makes sense. I just released GraphQL-Pro v1.24.13 which accepts `context: { ... }` in `AddOperationBatch.call`, want to give that a try? 

Another option might be to make a custom Context class for your schema, and set up some values in its `initialize` method: https://graphql-ruby.org/schema/definition.html#context-class

Let me know how it works for you!

**Comment by waissbluth** [comment]
Good news! Got it to work with the context. However, I needed to make a tweak to the code to get it to be passed in to `batch_upsert_client_operations` since that branch of the code does not seem to be covered.


Thanks for all the help getting to the bottom of this. It was not at all apparent to me that it was an app thing when we started.

Bad news is I moved on to the next error though 😅 , but that is probably out of scope for this issue. In any case, I am a bit surprised that this is failing both because a) I don't understand why this is showing up, and b) because it sounds more like a warning than an error?

```
mutation emailAddressDeletionMutation(
    $inputs: EmailAddressDeletionInputs!
) {
    emailAddressDeletion(inputs: $inputs) {
        destroyedId
    }
}
```

**Comment by rmosolgo** [comment]
Derp, I only implemented it for the Redis backend at first, sorry! I migrated my tests to cover both backends, added the patch you shared above, and released a new version, v1.24.14. Please give that version a try and let me know if you run into any other trouble! Thanks for your help working this out.
